### PR TITLE
Connected the UI to the backend using HTTP requests

### PR DESCRIPTION
**Pull request by davidardelean-debug**
1. Added HTTP requests
2. Created the checkout component
3. Implemented the order placement functionality 
4. Done style adjustments

### CODE REVIEW

File: src/components/cart-row/cart-row.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { useState } from "react";
+import { Link } from "react-router-dom";
+import { CartItem } from "../../entities/cart-item";
+import CartService from "../../services/cart-service";
+
+interface Props {
```
Comment (alexandra-gazda) [inline src/components/cart-row/cart-row.tsx]: CartRowProps would sound better. Please modify it for the other components too.

File: src/components/cart-row/cart-row.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  cart: CartItem[];
+  setCart: (newCart: CartItem[]) => void;
+}
+
+const CartRow = ({ cartItem, cart, setCart }: Props) => {
+  const { product, quantity } = cartItem;
+  const [newQuantity, setQuantity] = useState(quantity);
+
+  const handleQuantityChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const newQuantity = parseInt(event.target.value, 10);
```
Comment (alexandra-gazda) [inline src/components/cart-row/cart-row.tsx]: You can simply write const _newQuantity = parseInt(event.target.value)_ without radix, as it is an optional prop.

File: src/components/cart-row/cart-row.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    setCart(CartService.updateCartItemQuantity(cartItem, cart, newQuantity));
+  };
+  return (
+    <div className="cart-row">
+      <Link to={`/products/${product.id}`}><img src={product.imageUrl} width="30" height="30" /></Link>
+      <span className="product-name">{product.name}</span>
+      <span className="product-category">{product.category.name}</span>
+      <input
+        type="number"
+        id="product-quantity"
```
Comment (alexandra-gazda) [inline src/components/cart-row/cart-row.tsx]: It's not a really good practice to use ids (https://medium.com/@clairecodes/reasons-not-to-use-ids-in-css-a42204fb0d97) so maybe you can get rid of this one and use a class instead.

File: src/components/cart-row/cart-row.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        min="1"
+        value={newQuantity}
+        onChange={handleQuantityChange}
+      />
+      <span className="product-price">$ {product.price}</span>
+      <span className="product-subtotal">
+        $ {(product.price * quantity).toFixed(2)}
+      </span>
+      <button
+        className="btn remove-cart-item"
```
Comment (alexandra-gazda) [inline src/components/cart-row/cart-row.tsx]: In order to be more clear, rename this one with 'remove-cart-item-button' and in cart-row.module.scss just use the classname.

File: src/components/order-details/order-details.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { CartItem } from "../../entities/cart-item";
+
+interface Props{
+  cart: CartItem[]
+}
+const OrderDetails = ({cart}: Props) => {
+
+const total = cart.reduce((acc, cartItem)=> acc+= cartItem.product.price * cartItem.quantity ,0)
```
Comment (alexandra-gazda) [inline src/components/order-details/order-details.tsx]: Nice way to calculate the total :)

File: src/entities/cart-item.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import Product from "./product";
```
Comment (alexandra-gazda) [inline src/entities/cart-item.ts]: In order to be consistent, all the files in the 'entities' folder should be named with PascalCase (CartItem.ts). Also, please remove the commented line :D.

File: src/pages/cart/cart.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                cartItem={cartItem}
+                cart={cart}
+                setCart={setCart}
+              />
+            ))}
+            <div className="cart-totals">
+              Total:{" "}
+              <span className="sum">
+                $
+                {cart
```
Comment (alexandra-gazda) [inline src/pages/cart/cart.tsx]: You can extract this in a separate method in a _utils.ts_ file, as you calculate the total amount in order-datails too.

File: src/pages/product-listing/products-listing.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { useContext } from "react";
+import ProductListingItem from "../../components/product-listing-item/product-listing-item";
+import { ProductContext } from "../../providers/products-provider";
+
+const ProductsListing = () => {
+  const { contextProducts:products, error, isLoading } = useContext(ProductContext);
+  return (
+    <div>
```
Comment (alexandra-gazda) [inline src/pages/product-listing/products-listing.tsx:8]: You can avoid adding an extra node to the DOM and replace the <div> with the short syntax of React.Fragment. You should do this whenever possible.

File: src/pages/single-product/single-product.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+import { CartContext } from "../../providers/cart-provider";
+import { ProductContext } from "../../providers/products-provider";
+import APIClient from "../../services/api-client";
+import CartService from "../../services/cart-service";
+
+const SingleProduct = () => {
+  const { id } = useParams();
+  const navigate = useNavigate();
+
+  const { data: product, error, isLoading } = useProduct(id!);
```
Comment (alexandra-gazda) [inline src/pages/single-product/single-product.tsx:15]: I honestly don't like the non-null assertion (id!) and I would rather do a check before.
Check this out: https://blog.italanta.net/the-case-against-non-null-assertion-operator-ea6848aed636

File: src/pages/single-product/single-product.tsx [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    setQuantity(newQuantity);
+  };
+
+  const handleDelete = async () => {
+    if (confirm(`Are you sure you want to delete ${product?.name}?`)) {
+      await new APIClient(PRODUCTS_ENDPOINT).remove(id);
+      navigate("/products");
+      const newProducts = contextProducts.filter((item) => item.id !== id);
+      refetchProducts(newProducts);
+      const cartItem = cart.find((item) => item.product.id === id)!;
```
Comment (alexandra-gazda) [inline src/pages/single-product/single-product.tsx]: I honestly don't like the non-null assertion (!) and I would rather do a check before.
Check this out: https://blog.italanta.net/the-case-against-non-null-assertion-operator-ea6848aed636

File: src/entities/products-context-object.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import { SetStateAction } from "react";
+import Product from "./product";
+
+export interface ProductsContextObject {
+  contextProducts: Product[];
+  refetchProducts: (data: SetStateAction<Product[]>) => void;
```
Comment (alexandra-gazda) [inline src/entities/products-context-object.ts]: I really don't think that you should use refetchProducts. Is this one only used for the delete functionality?

File: src/services/cart-service.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  static removeFromCart(cartItem: CartItem, cart: CartItem[]): CartItem[] {
+    return cart.filter((item) => item.product.id != cartItem.product.id);
+  }
+
+  static updateCartItemQuantity(
+    cartItem: CartItem,
+    cart: CartItem[],
+    newQuantity: number,
+  ): CartItem[] {
+    const existingItem = cart.find(
```
Comment (alexandra-gazda) [inline src/services/cart-service.ts]: There is no need to extract the existingItem, you can write like this:

```
if (newQuantity != 0)
      return cart.map((item) =>
        item.product.id === cartItem.product.id
          ? { ...existingItem, quantity: newQuantity }
          : item,
      );
    else if ... 
```

File: src/components/cart-row/cart-row.tsx (hunk 3 above)
Comment (davidardelean-debug) [reply src/components/cart-row/cart-row.tsx]: Ok. 
I thought it is acceptable with respect to input elements inside a form. But I've used it for styling so you're right.

File: src/pages/product-listing/products-listing.tsx (hunk 8 above)
Comment (davidardelean-debug) [reply src/pages/product-listing/products-listing.tsx:8]: I kind of need the wrapper element in the way I designed the layout. I just forgot to assign it a class.
Besides that, all other components from `/pages` have a wrapper element for consistency.

File: src/pages/single-product/single-product.tsx (hunk 10 above)
Comment (davidardelean-debug) [reply src/pages/single-product/single-product.tsx]: Yes, indeed. In this scenario I wrongly assumed that a product that I am deleting is in the cart as well. Here a conditional statement is almost mandatory if using common sense.

However, I think there are scenarios where the non-null assertion may be useful, such as this case where I am retrieving the product id using `useParams()` hook inside the SingleProduct component, that is routed to a path containing an id.
As I cannot call hooks conditionally I must somehow guarantee to the compiler that the id is not null. (I am using `useEffect()` inside `useProduct()`).

image

P.S.: Really cool article, thank you.

File: src/entities/products-context-object.ts (hunk 11 above)
Comment (davidardelean-debug) [reply src/entities/products-context-object.ts]: For the moment yes, it is only used for the delete functionality. However, I may also use it for adding or updating products which I have to implement next.

I am aware that the current implementation of this is kind of "dirty", but I am sure things will get cleaner with a state management library.

I am open to suggestions and really curios about how I can improve this part.

File: src/services/cart-service.ts (hunk 12 above)
Comment (davidardelean-debug) [reply src/services/cart-service.ts]: Indeed, thank you for pointing it out!
You probably mean to use the spread operator on `cartItem` or `item.`
Here I think I duplicated the `addToCart` function and forgot to adjust the logic for the update function, only making essential changes.

**Pull request merged**